### PR TITLE
[stable/kube2iam] Support extra container environment variables

### LINKED ISSUE #5088: [stable/kube2iam] Support extra container environment variables, required for HTTP proxy
**Is this a request for help?**: 
No
---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):
FEATURE REQUEST

**Version of Helm and Kubernetes**:
`kubectl version`
```
Client Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.0", GitCommit:"fc32d2f3698e36b93322a3465f63a14e9f0eaead", GitTreeState:"clean", BuildDate:"2018-03-26T16:55:54Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"9+", GitVersion:"v1.9.2-eks.1", GitCommit:"e6c42d312ce5a80461ce10e6543e8fe0346bf065", GitTreeState:"clean", BuildDate:"2018-02-13T22:26:33Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"linux/amd64"}
```
`helm version`
```
Client: &version.Version{SemVer:"v2.9.0-rc3", GitCommit:"60abcdca41f544caaecb224acbfb92aee11e1f6e", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.9.0-rc3", GitCommit:"60abcdca41f544caaecb224acbfb92aee11e1f6e", GitTreeState:"clean"}
```

**Which chart**:
stable/kube2iam

**What happened**:
Communication from kube2iam to the AWS STS/IAM APIs needs to be routed via an HTTP proxy.
This is usually done by setting the `HTTP_PROXY` environment variable for a container but the kube2iam chart does not allow container environment variables to be set so the API call times out:
```
time="2018-04-16T20:45:08Z" level=error msg="Error assuming role RequestError: send request failed\ncaused by: Post https://sts.amazonaws.com/: dial tcp 54.239.29.25:443: i/o timeout" ns.name=default pod.iam.role="arn:aws:iam::1111111111111:role/RRRRRRRR" req.method=GET req.path="/latest/meta-data/iam/security-credentials/arn:aws:iam::1111111111111:role/RRRRRRRR" req.remote=x.x.x.x
```

**What you expected to happen**:
I expect to be able to specify additional environment variables for the container in the `values.yaml` or equivalent.

**How to reproduce it** (as minimally and precisely as possible):

**Anything else we need to know**:

### PR DESCRIPTION
**Pull request by ewbankkit**
**What this PR does / why we need it**:
Add support for extra container environment variables.
Useful when using an HTTP proxy for communication with AWS APIs.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #5088 

**Special notes for your reviewer**:
Some testing.
No extra environment variables:
```
helm install --dry-run --debug stable/kube2iam

COMPUTED VALUES:
affinity: {}
aws:
  access_key: ""
  region: ""
  secret_key: ""
extraArgs: {}
extraEnv: {}
host:
  interface: docker0
  ip: $(HOST_IP)
  iptables: false
  port: 8181
image:
  pullPolicy: IfNotPresent
  repository: jtblin/kube2iam
  tag: 0.10.0
nodeSelector: {}
podAnnotations: {}
podLabels: {}
probe: true
rbac:
  create: false
  serviceAccountName: default
resources: {}
tolerations: []
updateStrategy: OnDelete
verbose: false

HOOKS:
MANIFEST:

---
# Source: kube2iam/templates/daemonset.yaml
apiVersion: extensions/v1beta1
kind: DaemonSet
metadata:
  labels:
    app: kube2iam
    chart: kube2iam-0.8.7
    heritage: Tiller
    release: goodly-crab
  name: goodly-crab-kube2iam
spec:
  template:
    metadata:
      labels:
        app: kube2iam
        release: goodly-crab
    spec:
      containers:
        - name: kube2iam
          image: "jtblin/kube2iam:0.10.0"
          imagePullPolicy: "IfNotPresent"
          args:
            - --host-interface=docker0
            - --iptables=false
            - --app-port=8181
          env:
            - name: HOST_IP
              valueFrom:
                fieldRef:
                  fieldPath: status.podIP
          ports:
            - containerPort: 8181
          livenessProbe:
            httpGet:
              path: /healthz
              port: 8181
              scheme: HTTP
            initialDelaySeconds: 30
            periodSeconds: 5
            successThreshold: 1
            failureThreshold: 3
            timeoutSeconds: 1
          resources:
            {}
            
      hostNetwork: true
      serviceAccountName: "default"
      tolerations:
        []
        
  updateStrategy:
    type: OnDelete
```

Extra environment variables:
```
helm install --dry-run --debug stable/kube2iam

COMPUTED VALUES:
affinity: {}
aws:
  access_key: ""
  region: ""
  secret_key: ""
extraArgs: {}
extraEnv:
  NO_PROXY: localhost,127.0.0.1
  http_proxy: http://example.com:8080
host:
  interface: docker0
  ip: $(HOST_IP)
  iptables: false
  port: 8181
image:
  pullPolicy: IfNotPresent
  repository: jtblin/kube2iam
  tag: 0.10.0
nodeSelector: {}
podAnnotations: {}
podLabels: {}
probe: true
rbac:
  create: false
  serviceAccountName: default
resources: {}
tolerations: []
updateStrategy: OnDelete
verbose: false

HOOKS:
MANIFEST:

---
# Source: kube2iam/templates/daemonset.yaml
apiVersion: extensions/v1beta1
kind: DaemonSet
metadata:
  labels:
    app: kube2iam
    chart: kube2iam-0.8.7
    heritage: Tiller
    release: awesome-fly
  name: awesome-fly-kube2iam
spec:
  template:
    metadata:
      labels:
        app: kube2iam
        release: awesome-fly
    spec:
      containers:
        - name: kube2iam
          image: "jtblin/kube2iam:0.10.0"
          imagePullPolicy: "IfNotPresent"
          args:
            - --host-interface=docker0
            - --iptables=false
            - --app-port=8181
          env:
            - name: HOST_IP
              valueFrom:
                fieldRef:
                  fieldPath: status.podIP
            - name: NO_PROXY
              value: localhost,127.0.0.1
            - name: http_proxy
              value: http://example.com:8080
          ports:
            - containerPort: 8181
          livenessProbe:
            httpGet:
              path: /healthz
              port: 8181
              scheme: HTTP
            initialDelaySeconds: 30
            periodSeconds: 5
            successThreshold: 1
            failureThreshold: 3
            timeoutSeconds: 1
          resources:
            {}
            
      hostNetwork: true
      serviceAccountName: "default"
      tolerations:
        []
        
  updateStrategy:
    type: OnDelete
```

### CODE REVIEW

**Comment by ewbankkit** [comment]
Validated in my environment FWIW.

**Comment by ewbankkit** [comment]
Implementation is based on similar for Prometheus: https://github.com/kubernetes/charts/blob/d170629cc67d25c8528152348f027d64a4763414/stable/prometheus/templates/alertmanager-deployment.yaml#L35-L38

File: stable/kube2iam/templates/daemonset.yaml [hunk 1]
```diff
@@ -63,6 +63,10 @@ spec:
             - name: AWS_DEFAULT_REGION
               value: {{ .Values.aws.region }}
           {{- end }}
+          {{- range $name, $value := .Values.extraEnv }}
+            - name: {{ $name }}
+              value: {{ $value }}
```
Comment (unguiculus) [inline stable/kube2iam/templates/daemonset.yaml]: You should make sure the value is quoted.

**Review by unguiculus — CHANGES_REQUESTED** [review]

**Comment by ewbankkit** [comment]
Made requested review changes to quote environment variables values.

Testing:

```
helm install --dry-run --debug stable/kube2iam

COMPUTED VALUES:
affinity: {}
aws:
  access_key: ""
  region: ""
  secret_key: ""
extraArgs: {}
extraEnv:
  NO_PROXY: localhost,127.0.0.1
  http_proxy: http://example.com:8080
host:
  interface: docker0
  ip: $(HOST_IP)
  iptables: false
  port: 8181
image:
  pullPolicy: IfNotPresent
  repository: jtblin/kube2iam
  tag: 0.10.0
nodeSelector: {}
podAnnotations: {}
podLabels: {}
probe: true
rbac:
  create: false
  serviceAccountName: default
resources: {}
tolerations: []
updateStrategy: OnDelete
verbose: false

HOOKS:
MANIFEST:

---
# Source: kube2iam/templates/daemonset.yaml
apiVersion: extensions/v1beta1
kind: DaemonSet
metadata:
  labels:
    app: kube2iam
    chart: kube2iam-0.8.7
    heritage: Tiller
    release: cantankerous-horse
  name: cantankerous-horse-kube2iam
spec:
  template:
    metadata:
      labels:
        app: kube2iam
        release: cantankerous-horse
    spec:
      containers:
        - name: kube2iam
          image: "jtblin/kube2iam:0.10.0"
          imagePullPolicy: "IfNotPresent"
          args:
            - --host-interface=docker0
            - --iptables=false
            - --app-port=8181
          env:
            - name: HOST_IP
              valueFrom:
                fieldRef:
                  fieldPath: status.podIP
            - name: NO_PROXY
              value: "localhost,127.0.0.1"
            - name: http_proxy
              value: "http://example.com:8080"
          ports:
            - containerPort: 8181
          livenessProbe:
            httpGet:
              path: /healthz
              port: 8181
              scheme: HTTP
            initialDelaySeconds: 30
            periodSeconds: 5
            successThreshold: 1
            failureThreshold: 3
            timeoutSeconds: 1
          resources:
            {}
      hostNetwork: true
      serviceAccountName: "default"
      tolerations:
        []
  updateStrategy:
    type: OnDelete
```

**Review by unguiculus — APPROVED** [review]

**Pull request merged**